File: Python/Scripts/ex2_3_2.py
# exercise 2.3.2

from pylab import *
import scipy.linalg as linalg
from scipy.io import loadmat

# Digits to include in analysis (to include all, n = range(10) )
n = [0,1]
# Number of principal components for reconstruction
K = 16
# Digits to visualize
nD = range(6);


# Load Matlab data file to python dict structure
# and extract variables of interest
traindata = loadmat('../Data/zipdata.mat')['traindata']
X = mat(traindata[:,1:])
y = mat(traindata[:,0]).T

N,M = X.shape
C = len(n)

classValues = unique(y.A)
classNames = [str(int(v)) for v in classValues]
classDict = dict(zip(classNames,classValues))


# Select subset of digits classes to be inspected
class_mask = np.zeros(N).astype(bool)
for v in n:
    cmsk = y.A.ravel()==v
    class_mask = class_mask | cmsk
X = X[class_mask,:]
y = y[class_mask,:]
N=X.shape[0]

# Center the data (subtract mean column values)
Xc = X - np.ones((N,1))*X.mean(0)

# PCA by computing SVD of Y
U,S,V = linalg.svd(Xc,full_matrices=False)
U = mat(U)
V = mat(V).T

# Compute variance explained by principal components
rho = (S*S) / (S*S).sum() 

# Project data onto principal component space
Z = Xc * V

# Plot variance explained
figure()
plot(rho,'o-')
title('Variance explained by principal components');
xlabel('Principal component');
ylabel('Variance explained value');


# Plot PCA of the data
f = figure()
f.hold()
title('NanoNose data: PCA')
for c in n:
    # select indices belonging to class c:
    class_mask = y.A.ravel()==c
    plot(array(Z[class_mask,0]), array(Z[class_mask,1]), 'o')
legend(classNames)
xlabel('PC1')
ylabel('PC2')


# Visualize the reconstructed data from the first K principal components
# Select randomly D digits.
figure()
W = Z[:,range(K)] * V[:,range(K)].T
D = len(nD)
for d in range(D):
    digit_ix = np.random.randint(0,N)
    subplot(2, D, d+1)
    I = reshape(X[digit_ix,:], (16,16))
    imshow(I, cmap=cm.gray_r)
    title('Original')
    subplot(2, D, D+d+1)
    I = reshape(W[digit_ix,:]+X.mean(0), (16,16))
    imshow(I, cmap=cm.gray_r)
    title('Reconstr.');
    

# Visualize the pricipal components
figure()
for k in range(K):
    N1 = ceil(sqrt(K)); N2 = ceil(K/N1)
    subplot(N2, N1, k+1)
    I = reshape(V[:,k], (16,16))
    imshow(I, cmap=cm.hot)
    title('PC{0}'.format(k+1));

# output to screen
show()

File: Matlab/Scripts/ex2_3_2.m
%% exercise 2.3.2

% Digits to include in analysis (to include all, n = 1:10);
n = [0 1];

% Number of principal components for reconstruction
K = 12;

% Digits to visualize
nD = 1:5;

%% Load data
cdir = fileparts(mfilename('fullpath')); 
load(fullfile(cdir,'../Data/zipdata.mat'));

% Extract digits
X = traindata(:,2:end);
y = traindata(:,1);
classNames = {'0';'1';'2';'3';'4';'5';'6';'7';'8';'9';'10'};
classLabels = classNames(y+1);

% Remove digits that are not to be inspected
j = ismember(y, n);
X = X(j,:);
classLabels = classLabels(j);
classNames = classNames(n+1);
y = cellfun(@(str) find(strcmp(str, classNames)), classLabels)-1;

% Subtract the mean from the data
Y = bsxfun(@minus, X, mean(X));

% Obtain the PCA solution by calculate the SVD of Y
[U, S, V] = svd(Y, 'econ');

% Compute the projection onto the principal components
Z = U*S;

% Compute variance explained
rho = diag(S).^2./sum(diag(S).^2);

%% Plot variance explained
mfig('Digits: Var. explained'); clf;
plot(rho, 'o-');
title('Variance explained by principal components');
xlabel('Principal component');
ylabel('Variance explained value');

%% Plot PCA of data
mfig('Digits: PCA'); clf; hold all; 
C = length(classNames);
for c = 0:C-1
    plot(Z(y==c,1), Z(y==c,2), 'o');
end
legend(classNames);
xlabel('PC 1');
ylabel('PC 2');
title('PCA of digits data');

%% Visualize the reconstructed data from the firts K principal components
mfig('Digits: Reconstruction'); clf;
W = Z(:,1:K)*V(:,1:K)';
D = length(nD);
for d = 1:D
    subplot(2,D,d);
    I = reshape(X(nD(d),:), [16,16])';
    imagesc(I);
    axis image off
    title('Original');
    subplot(2,D,d+D);
    I = reshape(W(nD(d),:)+mean(X), [16,16])';
    imagesc(I);
    axis image off
    title('Reconstructed');
end
colormap(1-gray);

%% Visualize the pricipal components
mfig('Digits: Principal components'); clf;
for k = 1:K
    N1 = ceil(sqrt(K)); N2 = ceil(K/N1);
    subplot(N2, N1, k);
    W = U(:,k)*V(:,k)';
    I = reshape(W(k,:), [16,16])';
    imagesc(I);
    colormap(hot);
    axis image off
    title(sprintf('PC %d',k));
end
